### ISSUE TITLE
Filter issues

### ISSUE THREAD

**Issue by licaon-kter** [M66B/NetGuard#377]
Xperia 5.1.1, rooted, NetGuard 2.10 (no modification), no hosts blocked.

While testing for my other crash, I see [this](https://github.com/M66B/NetGuard/issues/291#issuecomment-197808851) more prominently.

Using the same setup, meaning apps (AFWall+/NetGuard/Conversations) excepted in battery optimisations ( see [1](https://i.imgur.com/7rAzaXX.jpg), [2](https://i.imgur.com/keCjTuJ.jpg); it's in *"apps active during standby"*) and just toggling `filter` on or off.

With `filter off` Conversations works ok (aka stays online/connected) when the screen is off, during both normal and on connection changes.

So I have some questions regarding `filter on` handling differences:
* what does `D NetGuard.JNI: pselect timeout` mean?
* what does `W NetGuard.JNI: pselect signaled tun 32 thread ae334000` mean?
* what's up with the numbers in `D NetGuard.JNI: pselect ready 1` or `2` or `3` or `4` ?
* `E NetGuard.JNI: connect error 101: Network is unreachable` or `E NetGuard.JNI: UDP sendto error 101: Network is unreachable` means that NetGuard just stopped handling the network?
* so next seeing `D conversations: account@host: ping timeout` and `D conversations: account@host/resourcename: time out during connect reconnecting` and `D conversations: account@host: failed to connect to /serveraddress (port ####) after 15000ms (java.net.SocketTimeoutException)` is normal?
* it's normal to see mostly `I NetGuard.Service: Messages=false wakelock=false` and rarely `true/true`

**Comment by licaon-kter** [comment]
Ok, even on screen on and unlocked `E NetGuard.JNI: connect error 101: Network is unreachable` or `E NetGuard.JNI: UDP sendto error 101: Network is unreachable` still appear ever other second.

**Comment by M66B** [comment]
To understand the logging, you can read the JNI source code. It is too much/too complicated to explain all. You can find the documentation for the called Linux functions by searching like this:

man pselect

Most of the error messages you want to know about can be found in these manual page as well.

**Comment by M66B** [comment]
ENETUNREACH = Network is unreachable

**Comment by M66B** [comment]
I like to see a more complete logcat of this (at least the VPN restarting because of screen on/off and the network unreachable error), because just a few highlighted log lines is not enough to see what is happening.

Meanwhile you can try this patch:

```
diff --git a/app/src/main/jni/netguard/netguard.c b/app/src/main/jni/netguard/netguard.c
index ce81470..4d53620 100644
--- a/app/src/main/jni/netguard/netguard.c
+++ b/app/src/main/jni/netguard/netguard.c
@@ -168,11 +168,10 @@ Java_eu_faircode_netguard_ServiceSinkhole_jni_1stop(
                 log_android(ANDROID_LOG_WARN, "pthread_join error %d: %s", err, strerror(err));
         }
 
-        if (clear) {
-            clear_icmp();
-            clear_udp();
+        clear_icmp();
+        clear_udp();
+        if (clear)
             clear_tcp();
-        }
 
         log_android(ANDROID_LOG_WARN, "Stopped thread %x", t);
     } else
```

My best guess this is all caused by routing errors of this Android version.
(which device was this?)

**Comment by M66B** [comment]
This commit wil stop stop/clear datagram (=ICMP/UDP) connections separately:
https://github.com/M66B/NetGuard/commit/ebdddb11efaf24a57f33936c661abba027b3dc83

I am closing this issue, because of "* ...even after a toggling off/on of the app*", which means that this probably cannot be solved from NetGuard.

Feel free to keep commenting here, though.

**Comment by M66B** [comment]
Relevant manual pages:

http://linux.die.net/man/2/pselect
http://linux.die.net/man/2/connect
http://linux.die.net/man/2/sendto
http://linux.die.net/man/2/recv

There are more, but these ones are the main ones you will need to understand the code / error messages.

**Comment by licaon-kter** [comment]
No comparative (filter on vs off) logcat today, but the day went like this: https://gist.github.com/licaon-kter/f767218ba04305a2c554

* before 10:00. phone was not used w/ `filter on`, reconnection timing went really high during the night
* 10:00 phone was used for a while
* before 11:40 I switched `filter off` since I needed Conversations to actually work
* forgot it like that until ~16:40, when I switched to `filter on` again
* before 19:00 I used it for a while
* ignored it again until 23:00 when I got the logcat

The only pattern that stands out is that it actually gets some pings out ( about 4 ) after screen off, since it does one at ~5min, that means ~20-25min. The battery optimisation thing might be a factor here (it has that `queue background data`), but it does not explain why with `filter off` it works continuously as expected.

**Comment by licaon-kter** [comment]
With `filter on`, today I test the [Stamina mode](https://i.imgur.com/keCjTuJ.jpg) influence, I disabled _"Extended standby"_ which greys out both _"Apps active in standby"_ and _"Stamina clock"_.

So far it looks fine, just like `filter off` did in the post above, like a clock, every ~5min a ping, not missing any messages.

**Comment by licaon-kter** [comment]
Got 2.12 since https://github.com/M66B/NetGuard/commit/eb99c51efdb404013dea494ec540dfd424fa32bf sounded like something that could help (Stamina yields `wakelock` related messages like `SuperStamina-WakelockHandler: unblock uids`), and so far (with [Stamina active again](https://i.imgur.com/keCjTuJ.jpg)) it's working like it should.

**Comment by licaon-kter** [comment]
Yes, looks like https://github.com/M66B/NetGuard/commit/eb99c51efdb404013dea494ec540dfd424fa32bf really helped, `filter on` works ok, at least concerning Conversations.

Now, I saw another issue with https://github.com/n76/Local-GSM-Backend, on database generation it will get stuck and will error out after a bit with `connection reset by peer`, retried a dozen times to no avail, switched `filter off` on a hunch and it worked right away. I'll see about a logcat later.

**Comment by M66B** [comment]
I don't believe this commit will change anything regarding this problem.

Try changing the TCP idle timeout in the native header file to a much larger value.


licaon-kter <notifications@github.com> schreef op 24 maart 2016 07:35:35 CET:

**Comment by licaon-kter** [comment]
> I don't believe this commit will change anything regarding this problem.

You mean for `reset by peer` or `Conversations disconnect on screen off`?

**Comment by M66B** [comment]
Neither of them

licaon-kter <notifications@github.com> schreef op 24 maart 2016 11:11:34 CET:

**Comment by licaon-kter** [comment]
Looks like I need to downgrade to 2.11 and retest then.

**Comment by licaon-kter** [comment]
Yes, of course you are right, back on 2.11 and works ok.
Might have been something else then.
